### PR TITLE
Fixed crashes and minor refactoring.

### PR DESCRIPTION
**Pull request by lorc**
Client library crashes if we try to work with unitialized session.
This patch adds checks on values returned by TEE_OpenSession() to
avoid segfaults.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

### CODE REVIEW

File: host/xtest/xtest_20000.c [hunk 1]
```diff
@@ -459,13 +460,15 @@ static void storage_corrupt(ADBG_Case_t *c,
...
 				TEE_DATA_FLAG_ACCESS_WRITE, 0, filedata,
-				tv->data_len, &obj_id));
+			        tv->data_len, &obj_id)))
 
-		ADBG_EXPECT(c, TEE_SUCCESS,
-			    obj_close(&sess, obj_id));
+			ADBG_EXPECT(c, TEE_SUCCESS,
+			            obj_close(&sess, obj_id));
+		else
+			goto exit;
```
Comment (igoropaniuk) [inline host/xtest/xtest_20000.c]: Possible memleak. By exiting here we don't free filedata (allocation is on line 451).
The same issue on lines 503/533

File: host/xtest/xtest_1000.c [hunk 2]
```diff
@@ -411,9 +411,10 @@ static void xtest_tee_test_invalid_mem_access(ADBG_Case_t *c, uint32_t n)
...
 
-	(void)ADBG_EXPECT_TEEC_SUCCESS(c,
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
 		xtest_teec_open_session(&session, &os_test_ta_uuid, NULL,
-					&ret_orig));
+		                        &ret_orig)))
+		return;
 
 	op.params[0].value.a = n;
 	op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT, TEEC_NONE, TEEC_NONE,
```
Comment (etienne-lms) [inline host/xtest/xtest_1000.c]: I know it's a bit out of the scope of this pr. Would you mind fixing the identation rigth below. The 2 calls have same args but you must read twice to understand it.

File: host/xtest/xtest_20000.c [hunk 3]
```diff
@@ -427,9 +427,10 @@ static void storage_corrupt(ADBG_Case_t *c,
 	ADBG_EXPECT(c, TEE_SUCCESS,
```
Comment (etienne-lms) [inline host/xtest/xtest_20000.c]: This need fixing also.
I have been back through the gp specs and open session on an invalid context is an programming error.
I guess there are a full bunch of `TEE_OpenSession()` and `TEE_XXXSharedMemory()` on potentially uninited context around.
Could you have a round after them? Or we can open another pull for those (no issue reported yet).

File: host/xtest/xtest_1000.c [hunk 4]
```diff
@@ -764,109 +767,64 @@ static void *cancellation_thread(void *arg)
...
-						    ret_orig);
-		TEEC_CloseSession(&session);
-	}
-	Do_ADBG_EndSubCase(c, "TEE Wait 0.5s");
-
 #ifdef USER_SPACE
-	Do_ADBG_BeginSubCase(c, "TEE Wait 2s cancel");
+	pthread_t thr;
+#endif
+	Do_ADBG_BeginSubCase(c, "%s", subcase);
```
Comment (etienne-lms) [inline host/xtest/xtest_1000.c]: minor: looks better with a empty line below local vars.

File: host/xtest/xtest_1000.c [hunk 5]
```diff
@@ -764,109 +767,64 @@ static void *cancellation_thread(void *arg)
...
-					   &ret_orig));
+			op.params[0].value.a = timeout;
+			op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT,
+			                                 TEEC_NONE,
+			                                 TEEC_NONE, TEEC_NONE);
+#ifdef USER_SPACE
+			if (cancel)
+				(void)ADBG_EXPECT(c, 0,
+				      pthread_create(&thr, NULL,
+				                     cancellation_thread, &op));
```
Comment (etienne-lms) [inline host/xtest/xtest_1000.c]: minor: could remove this empty line

File: host/xtest/xtest_1000.c [hunk 6]
```diff
@@ -764,109 +767,64 @@ static void *cancellation_thread(void *arg)
...
 
-		(void)ADBG_EXPECT_TEEC_ERROR_ORIGIN(c, TEEC_ORIGIN_TRUSTED_APP,
-						    ret_orig);
-		(void)ADBG_EXPECT(c, 0, pthread_join(thr, NULL));
-		TEEC_CloseSession(&session);
-	}
-	Do_ADBG_EndSubCase(c, "TEE Wait 2s cancel");
 #endif
```
Comment (etienne-lms) [inline host/xtest/xtest_1000.c]: minor: could remove this empty line

File: host/xtest/xtest_1000.c [hunk 7]
```diff
@@ -764,109 +767,64 @@ static void *cancellation_thread(void *arg)
...
+			TEEC_CloseSession(&session);
+		}
+	}
+	Do_ADBG_EndSubCase(c, "%s", subcase);
 
-		op.params[0].value.a = 2000;
-		op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT, TEEC_NONE,
-						 TEEC_NONE, TEEC_NONE);
+}
+static void xtest_tee_test_1009(ADBG_Case_t *c)
```
Comment (etienne-lms) [inline host/xtest/xtest_1000.c]: minor: empty line

File: host/xtest/xtest_1000.c (hunk 2 above)
Comment (etienne-lms) [inline host/xtest/xtest_1000.c]: thanks

File: host/xtest/xtest_1000.c [hunk 8]
```diff
@@ -764,109 +767,64 @@ static void *cancellation_thread(void *arg)
...
+{
+	xtest_tee_test_1009_subcase(c, "TEE Wait 0.1s", 100, false);
+	xtest_tee_test_1009_subcase(c, "TEE Wait 0.5s", 500, false);
 
-		(void)ADBG_EXPECT_TEEC_SUCCESS(c,
-			TEEC_InvokeCommand(&session, TA_OS_TEST_CMD_WAIT, &op,
-					   &ret_orig));
+#ifdef USER_SPACE
```
Comment (etienne-lms) [inline host/xtest/xtest_1000.c]: minor: you can remove the empty lines around this subcase.

**Comment by lorc** [comment]
@etienne-lms 
Oops, looks like I can't respond to your code comments. Anyways, I tried to address all your comments. I I hope that code looks better now :)

**Comment by lorc** [comment]
@etienne-lms Yes, I was `xtest_5000.c`. I think it should be fixed in another patch. They are not included in standard test suite. Honestly, I don't know how to run them. Looks like I need GP Test Suite?

**Comment by etienne-lms** [comment]
@lorc, thanks for the work. looks ok to me.

for xtest_5000, there is only 1 test remaining: 5006. It collects calls of the Client API with various arguments to check compliance. It is part of the generic xtest.
anyway, we can deal with it in another pull.

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by igoropaniuk** [comment]
`Reviewed-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`

**Comment by lorc** [comment]
Updated commit message.

**Comment by jforissier** [comment]
Thanks @lorc, could you please rebase on top of the current master branch please?

**Comment by lorc** [comment]
@jforissier, done

**Pull request merged**